authorization/invitation/model/invitation_model_service.go

```diff
@@ -0,0 +1,201 @@
+		inviteToIdentity, err = repo.Identities().Load(ctx, inviteToUUID)
+		if err != nil {
+			// That didn't work, try to lookup a resource with the same ID value
+			inviteToResource, err = repo.ResourceRepository().Load(ctx, inviteTo)
+			if err != nil {
+				return errors.NewNotFoundError(fmt.Sprintf("invalid identifier '%s' provided for organization, team, security group or resource", inviteTo), inviteTo)
+			}
+		}
+	}
+
+	if inviteToIdentity != nil {
+		// Load the resource for the identity
+		identityResource, err = repo.ResourceRepository().Load(ctx, *inviteToIdentity.IdentityResourceID)
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "error loading resource for identity")
+		}
+
+		// Confirm that the issuing user has the necessary scope to manage members for the organization, team or security group
+		scope, err := s.permModelSvc.HasScope(ctx, issuingUserId, *inviteToIdentity.IdentityResourceID, authorization.ManageMembersScope)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
+		}
+
+		// We only allow membership in some identity types - confirm that we are inviting to an organization, team or security group
+		if !authorization.CanHaveMembers(identityResource.ResourceType.Name) {
+			return errors.NewInternalErrorFromString(ctx, "may only invite a user to an organization, team or security group")
+		}
+	} else if inviteToResource != nil {
+		// Confirm that the issuing user has the manage members scope for the resource
+		scope, err := s.permModelSvc.HasScope(ctx, issuingUserId, inviteToResource.ResourceID, authorization.ManageMembersScope)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
+		}
+	}
+
+	// Iterate through all of the invitations and confirm that for each one:
+	// 1) a valid user has been specified via its User ID, e-mail address or username
+	// 2) any roles specified are valid roles for the organization, team or security group
+	// For each invitation, ensure that the IdentityID value can be found and set it
+	for i, invitation := range invitations {
+		// If the UserID has been provided, confirm it is valid and that the identity is a user
+		if invitation.IdentityID != nil {
+			identity, err := repo.Identities().Load(ctx, *invitation.IdentityID)
+			if err != nil {
+				return errors.NewInternalErrorFromString(ctx, fmt.Sprintf("invalid identity ID specified: %s", invitation.IdentityID))
+			}
+
+			if !identity.IsUser() {
+				return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
+			}
+		} else if invitation.UserName != nil {
```

Comment:
From my previous review:
> We should not do any look up by usernames or emails here. Our invitation API should accept identity(user) IDs only. In UI we will search for user anyway. So UI will know the identity ID. We don't have to do this search twice.

So, why do you think our API should support invite by usernames or email? Why not Identity ID only?

Reply:
After our discussion I've removed support for username and e-mail.